WalletWasabi/Cache/IdempotencyRequestCache.cs

```diff
@@ -50,14 +44,11 @@ public IdempotencyRequestCache(IMemoryCache cache)
 		bool callAction = false;
 		TaskCompletionSource<TResponse>? responseTcs;
 
-		lock (ResponseCacheLock)
+		if (!ResponseCache.TryGetValue(request, out responseTcs))
 		{
-			if (!ResponseCache.TryGetValue(request, out responseTcs))
-			{
-				callAction = true;
-				responseTcs = new();
-				ResponseCache.Set(request, responseTcs, options);
-			}
+			callAction = true;
```

Comment:
> I would love to understand the reasoning behind removing the lock which makes the algorithm correct.

What we currently have is just an illusion of correctness. Each controller has its own instance of `IdempotencyRequestCache `   and the controllers' lifetime is trassient, what means that for each request a new instance of the controller is created and each of those controllers create a new instance of the  `IdempotencyRequestCache `  so, there is not shared resource to synchronize. The only share resource is the `IMemoryCache` but the access to that has never be synchronized.

So, what does this `lock` do then? `lock(obj)`  statements give us exclusive access to the object `obj` that in this case, given it is an instance that is never shared, would give us exclusive access to something that nobody else have access anyway.

> Are you sure that this is like the root cause of issues?

No. I don't think this is the root cause at all but it doesn't make the things easier either because it is wrong and makes things confusing. Basically, removing this doesn't change nothing for worse, only for better.

> Cache JSON strings (as well) to avoid computing hash codes of requests. This would probably save a lot of CPU resources.
(Server static content without actually hitting ASP, filters might fall into this category)

These two alternatives are in my radar, yes.

Reply:
I've checked again an it seems `WabiSabiController` uses a singleton instance while `BlockchainController` and `SmartBlockProvider` create their own instance. What a mess, devs are using the `IdempontencyRequestCache` for GETs (which should be idempotent by definition) instead of only for POSTs. I will revert this I try to fix all this.

Reply:
> Does anyone have some stats regarding request numbers on the Backend server? I talked about this once with @kristapsk but I believe it was before we had the monitoring set up..

![image](https://github.com/zkSNACKs/WalletWasabi/assets/4500994/df9f48a3-4f5e-4f4e-954c-1fcb90276d40)